File: android/liverelay-sdk/src/main/java/app/orizn/liverelay/PeerConnectionProvider.kt
package app.orizn.liverelay

import android.content.Context
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.delay
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.MediaConstraints
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.SdpObserver
import org.webrtc.SessionDescription
import org.webrtc.audio.JavaAudioDeviceModule

/**
 * Centralised WebRTC bootstrap: a single [PeerConnectionFactory] and a shared
 * [EglBase] context used for capture, encoding and rendering.
 *
 * Mirrors the JS SDK flow (static/liverelay.js v0.5.0):
 * createOffer -> tuneOpusSdp/preferH264 (SDP munging) -> setLocalDescription
 * -> wait for ICE gathering (capped at 2s, partial candidates are fine because
 * the SFU sits on a public IP) -> POST offer / setRemoteDescription(answer).
 *
 * Note: SdpObserver callbacks fire on internal WebRTC threads; the suspend
 * helpers below simply resume the calling coroutine from those threads.
 */
object PeerConnectionProvider {

    /** Default cap on ICE gathering, matching waitForIceGathering() in liverelay.js. */
    private const val ICE_GATHERING_TIMEOUT_MS = 2_000L
    private const val ICE_POLL_INTERVAL_MS = 50L

    private val initialized = AtomicBoolean(false)

    @Volatile
    private var appContext: Context? = null

    /** Shared EGL context for capture, hardware encode and rendering. */
    val eglBase: EglBase by lazy { EglBase.create() }

    /**
     * Shared factory. Hardware-accelerated H.264 (high profile enabled) via
     * [DefaultVideoEncoderFactory], default [JavaAudioDeviceModule] for audio.
     * Requires [initialize] to have been called first.
     */
    val factory: PeerConnectionFactory by lazy {
        val context = appContext
            ?: throw LiveRelayException("PeerConnectionProvider.initialize(context) must be called before accessing factory")
        val encoderFactory = DefaultVideoEncoderFactory(
            eglBase.eglBaseContext,
            /* enableIntelVp8Encoder = */ true,
            /* enableH264HighProfile = */ true
        )
        val decoderFactory = DefaultVideoDecoderFactory(eglBase.eglBaseContext)
        val audioDeviceModule = JavaAudioDeviceModule.builder(context)
            .createAudioDeviceModule()
        PeerConnectionFactory.builder()
            .setVideoEncoderFactory(encoderFactory)
            .setVideoDecoderFactory(decoderFactory)
            .setAudioDeviceModule(audioDeviceModule)
            .createPeerConnectionFactory()
    }

    /**
     * Loads native libraries and initialises the global WebRTC state.
     * Idempotent: subsequent calls are no-ops.
     */
    fun initialize(context: Context) {
        if (!initialized.compareAndSet(false, true)) return
        val application = context.applicationContext
        appContext = application
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(application)
                .createInitializationOptions()
        )
    }

    /**
     * Creates a [PeerConnection] configured for the LiveRelay SFU
     * (Unified Plan, continual gathering, bundled + muxed transport).
     */
    fun makePeerConnection(
        iceServers: List<IceServerDto>,
        observer: PeerConnection.Observer
    ): PeerConnection {
        val rtcIceServers = iceServers.map { dto ->
            PeerConnection.IceServer.builder(dto.urls)
                .apply {
                    dto.username?.let { setUsername(it) }
                    dto.credential?.let { setPassword(it) }
                }
                .createIceServer()
        }
        val config = PeerConnection.RTCConfiguration(rtcIceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
            continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
            bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE
            rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE
            iceCandidatePoolSize = 2
        }
        return factory.createPeerConnection(config, observer)
            ?: throw LiveRelayException("PeerConnectionFactory.createPeerConnection returned null")
    }

    /**
     * createOffer -> low-latency SDP munging (Opus minptime/FEC, H.264 first)
     * -> setLocalDescription -> wait for ICE gathering (capped at 2000ms;
     * resolves with partial candidates on timeout, never fails on it).
     */
    suspend fun makeTunedOffer(pc: PeerConnection): SdpPayload {
        val offer = createOffer(pc)
        val tunedSdp = LatencyTuning.preferH264(LatencyTuning.tuneOpusSdp(offer.description))
        val tunedOffer = SessionDescription(SessionDescription.Type.OFFER, tunedSdp)
        setLocalDescription(pc, tunedOffer)
        awaitIceGathering(pc, ICE_GATHERING_TIMEOUT_MS)
        val local = pc.localDescription
            ?: throw LiveRelayException("localDescription is null after setLocalDescription")
        return SdpPayload(sdp = local.description, type = "offer")
    }

    /** Applies the SFU's answer to the connection. */
    suspend fun setRemoteAnswer(pc: PeerConnection, answer: SdpPayload) {
        val description = SessionDescription(SessionDescription.Type.ANSWER, answer.sdp)
        setRemoteDescription(pc, description)
    }

    // ------------------------------------------------------------------
    // Internals
    // ------------------------------------------------------------------

    private suspend fun createOffer(pc: PeerConnection): SessionDescription =
        suspendCancellableCoroutine { continuation ->
            pc.createOffer(object : SdpObserver {
                override fun onCreateSuccess(description: SessionDescription?) {
                    if (description != null) {
                        continuation.resume(description)
                    } else {
                        continuation.resumeWithException(
                            LiveRelayException("createOffer returned null description")
                        )
                    }
                }

                override fun onCreateFailure(error: String?) {
                    continuation.resumeWithException(
                        LiveRelayException("createOffer failed: ${error ?: "unknown error"}")
                    )
                }

                override fun onSetSuccess() { /* not used for createOffer */ }
                override fun onSetFailure(error: String?) { /* not used for createOffer */ }
            }, MediaConstraints())
        }

    private suspend fun setLocalDescription(pc: PeerConnection, description: SessionDescription) =
        suspendCancellableCoroutine { continuation ->
            pc.setLocalDescription(object : SdpObserver {
                override fun onSetSuccess() {
                    continuation.resume(Unit)
                }

                override fun onSetFailure(error: String?) {
                    continuation.resumeWithException(
                        LiveRelayException("setLocalDescription failed: ${error ?: "unknown error"}")
                    )
                }

                override fun onCreateSuccess(sdp: SessionDescription?) { /* not used for set */ }
                override fun onCreateFailure(error: String?) { /* not used for set */ }
            }, description)
        }

    private suspend fun setRemoteDescription(pc: PeerConnection, description: SessionDescription) =
        suspendCancellableCoroutine { continuation ->
            pc.setRemoteDescription(object : SdpObserver {
                override fun onSetSuccess() {
                    continuation.resume(Unit)
                }

                override fun onSetFailure(error: String?) {
                    continuation.resumeWithException(
                        LiveRelayException("setRemoteDescription failed: ${error ?: "unknown error"}")
                    )
                }

                override fun onCreateSuccess(sdp: SessionDescription?) { /* not used for set */ }
                override fun onCreateFailure(error: String?) { /* not used for set */ }
            }, description)
        }

    /**
     * Waits until [PeerConnection.iceGatheringState] reaches COMPLETE, polling
     * lightly. On timeout it returns normally so the caller proceeds with the
     * candidates gathered so far (same semantics as the JS SDK).
     */
    private suspend fun awaitIceGathering(pc: PeerConnection, timeoutMs: Long) {
        withTimeoutOrNull(timeoutMs) {
            while (pc.iceGatheringState() != PeerConnection.IceGatheringState.COMPLETE) {
                delay(ICE_POLL_INTERVAL_MS)
            }
        }
    }
}

File: ios/Sources/LiveRelaySDK/PeerConnectionProvider.swift
import Foundation
import WebRTC

/// Factory + helpers for RTCPeerConnection, mirroring the web SDK flow
/// (see WEB_RTC/static/liverelay.js): createOffer → tuneOpusSdp →
/// setLocalDescription → wait for ICE gathering (cap 2s) → POST offer.
public final class PeerConnectionProvider {

    public static let shared = PeerConnectionProvider()

    /// Single SSL init for the whole process.
    private static let sslInitialized: Bool = {
        RTCInitializeSSL()
        return true
    }()

    /// Shared factory with hardware H.264 (VideoToolbox) via the default
    /// encoder/decoder factories.
    public let factory: RTCPeerConnectionFactory

    private init() {
        _ = PeerConnectionProvider.sslInitialized
        let encoderFactory = RTCDefaultVideoEncoderFactory()
        let decoderFactory = RTCDefaultVideoDecoderFactory()
        factory = RTCPeerConnectionFactory(
            encoderFactory: encoderFactory,
            decoderFactory: decoderFactory
        )
    }

    // MARK: - Peer connection creation

    public func makePeerConnection(
        iceServers: [IceServerDTO],
        delegate: RTCPeerConnectionDelegate?
    ) -> RTCPeerConnection {
        let config = RTCConfiguration()
        config.iceServers = iceServers.map { dto in
            if let username = dto.username, let credential = dto.credential {
                return RTCIceServer(
                    urlStrings: dto.urls,
                    username: username,
                    credential: credential
                )
            }
            return RTCIceServer(urlStrings: dto.urls)
        }
        config.sdpSemantics = .unifiedPlan
        config.continualGatheringPolicy = .gatherContinually
        config.iceCandidatePoolSize = 2
        config.bundlePolicy = .maxBundle
        config.rtcpMuxPolicy = .require

        let constraints = RTCMediaConstraints(
            mandatoryConstraints: nil,
            optionalConstraints: nil
        )

        guard let pc = factory.peerConnection(
            with: config,
            constraints: constraints,
            delegate: delegate
        ) else {
            // Only fails on invalid configuration, which cannot happen with
            // the hardcoded values above.
            fatalError("RTCPeerConnectionFactory failed to create a peer connection")
        }
        return pc
    }

    // MARK: - Offer / answer

    /// createOffer → LatencyTuning.tuneOpusSdp → setLocalDescription →
    /// wait for ICE gathering completion (capped at 2s, resolves with the
    /// candidates gathered so far on timeout — the SFU sits on a public IP).
    public func makeTunedOffer(for pc: RTCPeerConnection) async throws -> SdpPayload {
        let constraints = RTCMediaConstraints(
            mandatoryConstraints: nil,
            optionalConstraints: nil
        )

        // 1. createOffer
        let offer: RTCSessionDescription = try await withCheckedThrowingContinuation { continuation in
            pc.offer(for: constraints) { sdp, error in
                if let error = error {
                    continuation.resume(throwing: LiveRelayError.webrtc("createOffer failed: \(error.localizedDescription)"))
                } else if let sdp = sdp {
                    continuation.resume(returning: sdp)
                } else {
                    continuation.resume(throwing: LiveRelayError.webrtc("createOffer returned no SDP"))
                }
            }
        }

        // 2. Low-latency Opus munging.
        let tunedSdp = LatencyTuning.tuneOpusSdp(offer.sdp)
        let tunedOffer = RTCSessionDescription(type: .offer, sdp: tunedSdp)

        // 3. setLocalDescription
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            pc.setLocalDescription(tunedOffer) { error in
                if let error = error {
                    continuation.resume(throwing: LiveRelayError.webrtc("setLocalDescription failed: \(error.localizedDescription)"))
                } else {
                    continuation.resume(returning: ())
                }
            }
        }

        // 4. Wait for ICE gathering (cap 2s, never throws on timeout).
        await waitForIceGathering(pc, timeoutSeconds: 2.0)

        // 5. Return the final local description (includes gathered candidates).
        let finalSdp = pc.localDescription?.sdp ?? tunedSdp
        return SdpPayload(sdp: finalSdp, type: "offer")
    }

    public func setRemoteAnswer(_ answer: SdpPayload, on pc: RTCPeerConnection) async throws {
        let remote = RTCSessionDescription(type: .answer, sdp: answer.sdp)
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            pc.setRemoteDescription(remote) { error in
                if let error = error {
                    continuation.resume(throwing: LiveRelayError.webrtc("setRemoteDescription failed: \(error.localizedDescription)"))
                } else {
                    continuation.resume(returning: ())
                }
            }
        }
    }

    // MARK: - ICE gathering wait

    /// Polls `iceGatheringState` until `.complete` or the timeout elapses.
    /// Always resolves (mirrors `waitForIceGathering` in liverelay.js):
    /// partial candidates are sufficient because the SFU has a public IP.
    private func waitForIceGathering(_ pc: RTCPeerConnection, timeoutSeconds: Double) async {
        let pollIntervalNs: UInt64 = 50_000_000 // 50 ms
        let deadline = Date().addingTimeInterval(timeoutSeconds)
        while pc.iceGatheringState != .complete && Date() < deadline {
            do {
                try await Task.sleep(nanoseconds: pollIntervalNs)
            } catch {
                return // Task cancelled — proceed with what we have.
            }
        }
    }
}
